### PR TITLE
Adding support for importing simple glTF files

### PR DESCRIPTION
**Pull request by xtanion**
This PR will allow us to load gltf files and get actors or `polydata` from it.

It currently supports:
- [x] multiple Nodes
- [x] multiple Meshes
- [x] Normals 
- [x] Cameras
- [x] Textures (baseColor)

Demo: 
image

 

WIP: glTF exporter, tests

### CODE REVIEW

**Comment by m-agour** [comment]
Hi @xtanion,
This looks good and well structured. The little duck looks so cute!
There are some issues that I found:

1. `var: list[int]` will cause the error `TypeError: 'type' object is not subscriptable` since the ability to use the `[]` operator on types like `list` was added in 3.9+. So you may add `from __future__ import annotations` at the first import or implement it any other supported way.
2. For Python 3.9 I got the following error:
    `ValueError: mutable default <class 'list'> for field nodes is not allowed: use default_factory` which is self explanatory.

I will address each line of code relate to these errors so you know what caused which error.

File: fury/gltf.py [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+@dataclass
+class Buffer:
+    byteLength: int
+    uri: str = None
+    name: str = None
+
+
+@dataclass
+class Scene:
+    nodes: list[int] = [0]
```
Comment (m-agour) [inline fury/gltf.py:134]: use `default_factory` as suggested by the error `ValueError: mutable default <class 'list'> for field nodes is not allowed: use default_factory
`

File: fury/gltf.py [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+    name: str = None
+
+
+@dataclass
+class Material:
+    pbrMetallicRoughness: dict = None
+    occlusionTexture: dict = None
+    normalTexture: dict = None
+    emissiveTexture: dict = None
+    emissiveFactor: list[int] = None
```
Comment (m-agour) [inline fury/gltf.py:53]: Only works for python 3.9+

File: fury/gltf.py [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+
+@dataclass
+class Scene:
+    nodes: list[int] = [0]
+
+
+@dataclass
+class glTF:
+    scene: int
+    scenes: list[Scene]
```
Comment (m-agour) [inline fury/gltf.py:140]: Only works in python 3.9+

File: fury/gltf.py [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+    accessors: list[Accessor] = None
```
Comment (m-agour) [inline fury/gltf.py:141]: Only works in python 3.9+

File: fury/gltf.py [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+    animations: dict = None
+    asset: dict = None
+    bufferViews: list[BufferView] = None
```
Comment (m-agour) [inline fury/gltf.py:144]: Only works in python 3.9+

File: fury/gltf.py [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+    buffers: list[Buffer] = None
```
Comment (m-agour) [inline fury/gltf.py:145]: Only works in python 3.9+

File: fury/gltf.py [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+    cameras: list[Camera] = None
```
Comment (m-agour) [inline fury/gltf.py:146]: Only works in python 3.9+

File: fury/gltf.py [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+    images: dict = None
+    materials: list[Material] = None
```
Comment (m-agour) [inline fury/gltf.py:148]: Only works in python 3.9+

File: fury/gltf.py [hunk 9]
```diff
@@ -0,0 +1,398 @@
...
+    meshes: list[Mesh] = None
```
Comment (m-agour) [inline fury/gltf.py:149]: Only works in python 3.9+

File: fury/gltf.py [hunk 10]
```diff
@@ -0,0 +1,398 @@
...
+    nodes: list[Node] = None
```
Comment (m-agour) [inline fury/gltf.py:150]: Only works in python 3.9+

File: fury/transform.py [hunk 11]
```diff
@@ -203,3 +204,139 @@ def cart2sphere(x, y, z):
...
+    return scale_mat
+
+
+def apply_transfomation(vertices, transformation):
+    """ Multiplying transformation matrix with vertices
+
+    Parameters
+    ------------
+    vertices : ndarray (n, 3)
+        verices of the mesh
```
Comment (m-agour) [inline fury/transform.py:326]: Typo in vertices

File: fury/gltf.py (hunk 1 above)
Comment (xtanion) [reply fury/gltf.py:134]: Thanks for the wonderful review @m-agour, I didn't know it was a python>=3.9 feature. I will look into it.

**Comment by xtanion** [comment]
Hi @m-agour , I modified `gltf.py` to use `typing.List` instead of python list. I tried running it on a `python3.6.10` venv and It worked fine. Can you check if it's working correctly on your system?
Thank You.

**Review by ganimtron-10 — COMMENTED** [review]
Hello @xtanion ,
That duck model looks cool!
Before the current changes, I was unable to run the tutorial but now it works fine.

Also when I tried to import another model it is throwing the below error

Model links: 
- https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/CesiumMilkTruck/glTF-Embedded/CesiumMilkTruck.gltf
- https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/2CylinderEngine/glTF-Embedded/2CylinderEngine.gltf

Error:
`json.decoder.JSONDecodeError: Expecting value: line 8 column 1 (char 7)`

Thanks!

File: fury/gltf.py (hunk 1 above)
Comment (ganimtron-10) [reply fury/gltf.py]: +1 Getting the same error on Python 3.9.13

**Comment by xtanion** [comment]
Hi @ganimtron-10, the problem with these links is they are not saved as a gltf file. if you open `CesiumMilkTruck.gltf` file you'll see that it stores the HTML of the page, not the actual `.gltf` file. You can try `raw` version of the links (for eg: [cesiummiltruck](https://raw.githubusercontent.com/KhronosGroup/glTF-Sample-Models/master/2.0/CesiumMilkTruck/glTF-Embedded/CesiumMilkTruck.gltf)) and they should work just fine.

**Comment by m-agour** [comment]
> Hi @m-agour , I modified `gltf.py` to use `typing.List` instead of python list. I tried running it on a `python3.6.10` venv and it worked fine. Can you check if it's working correctly on your system? Thank You.

Hi @xtanion,
I got an error related to the Duck.gltf file `No such file or directory: 'local-glTF/glTF-samples/duck/Duck.gltf'`, but I added the `duck.gltf` manually, and everything worked just fine.

**Comment by m-agour** [comment]
Hello @xtanion,
I have another suggestion regarding how you loaded the glTF model. I suggest you make a PR and add some glTF models to [fury-data](https://github.com/fury-gl/fury-data)  repo (there are already obj models there maybe make a glTF directory). Then update `fury.fetcher` to be able to fetch the glTF models.

File: fury/gltf.py (hunk 1 above)
Comment (Nibba2018) [reply fury/gltf.py]: Hi @m-agour and @ganimtron-10 , inorder to type hint lists you need to do the following:
```python
from typing import List
nodes: List[int] = [0]
```

Comment (Nibba2018) [reply fury/gltf.py]: Works on all existing python versions

File: docs/tutorials/01_introductory/viz_gltf.py [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+filename = os.path.basename(filename[len(filename)-1])
+
+urlretrieve(url, filename=filename)
+
+##############################################################################
+# Initialize the glTFImporter and get actors using `get_actors` method.
+# Note: You can always manulally create actor from polydata, and apply texture
+# or materials manually afterwards.
+
+importer = glTF(filename, apply_normals=True)
```
Comment (skoudoro) [inline docs/tutorials/01_introductory/viz_gltf.py:33]: gltf or gltf_obj instead of `importer`

**Review by skoudoro — CHANGES_REQUESTED** [review]
Hi @xtanion,

before going further into the review. Can you:
- update the PR with your last version
- merge master and add the fetcher 
- document all functions
- fix all pep8 issues
- make sure that the tests cover 90% of your code.
- See my comments below for your tutorial.
- use `load_image` and check a simpler way to flip the image

Please, let us know when it is ready. Thank you

File: fury/utils.py [hunk 13]
```diff
@@ -403,6 +403,27 @@ def get_polydata_vertices(polydata):
     return numpy_support.vtk_to_numpy(polydata.GetPoints().GetData())
 
 
+def get_polydata_tcoord(polydata):
```
Comment (skoudoro) [inline fury/utils.py:406]: why you do not use it? need a test

File: requirements/default.txt [hunk 14]
```diff
@@ -6,3 +6,4 @@ pillow>=5.4.1
 packaging
 Pillow>=8.0.1
 aiohttp>=3.7.4
+pygltflib
```
Comment (skoudoro) [inline requirements/default.txt:9]: please, add a minimal version

**Review by skoudoro — COMMENTED** [review]
Hi @xtanion,

See below for some comments. I still need to test this PR.

File: fury/transform.py [hunk 15]
```diff
@@ -203,3 +204,140 @@ def cart2sphere(x, y, z):
     phi = np.arctan2(y, x)
     r, theta, phi = np.broadcast_arrays(r, theta, phi)
     return r, theta, phi
+
+
+def translate(translation):
+    """Return transformation matrix for translation array.
+
+    Parameters
+    ------------
```
Comment (skoudoro) [inline fury/transform.py:213]: too long

File: fury/gltf.py [hunk 16]
```diff
@@ -0,0 +1,348 @@
...
+                utils.set_polydata_triangles(polydata, indices)
+
+            material = None
+            if primitive.material is not None:
+                material = self.get_materials(primitive.material)
+
+            self.polydatas.append(polydata)
+            self.materials.append(material)
+
+    def get_acc_data(self, acc_id):
```
[the remark this answers is not in the document]
Comment (xtanion) [reply fury/gltf.py:178]: I wrote `acc` as a short form of `accessor`. Some lines can go long if I write the entire word.

File: fury/gltf.py [hunk 17]
```diff
@@ -0,0 +1,348 @@
...
+        """Gets the textures data
+
+        Parameters
+        ----------
+        mat_id : int
+            Material index
+        """
+
+        material = self.gltf.materials[mat_id]
+        bct, mrt, nt = None, None, None
```
[the remark this answers is not in the document]
Comment (xtanion) [reply fury/gltf.py:277]: As of now, we don't need them, I'll remove them for now.

**Comment by ganimtron-10** [comment]
Hello @xtanion ,
While I was trying to execute the `viz_gltf` file, I am facing some strange issues mentioned below.
- The fetcher is giving `RuntimeError: Event loop is closed` which will get solved when you merge the #616 as we discussed.
- Also after successfully downloading the Model while importing, I am getting the below error
```python
Failed to read ! Error in opening file
Traceback (most recent call last):
  File "\docs\tutorials\01_introductory\viz_gltf.py", line 28, in <module>
    gltf_obj = glTF(filename, apply_normals=True)
  File "\fury\gltf.py", line 50, in __init__
    self.get_nodes(0)
  File "\fury\gltf.py", line 83, in get_nodes
    self.transverse_node(node_id, self.init_transform)
  File "\fury\gltf.py", line 129, in transverse_node
    self.transverse_node(child_id, next_matrix)
  File "\fury\gltf.py", line 121, in transverse_node
    self.load_mesh(mesh_id, next_matrix)
  File "\fury\gltf.py", line 148, in load_mesh
    vertices = transform.apply_transfomation(vertices, transform_mat)
  File "\fury\transform.py", line 335, in apply_transfomation
    shape = vertices.shape
AttributeError: 'NoneType' object has no attribute 'shape'
```
The vertices in the `apply_transformation` method is getting None.
Also as you mentioned I tried redownloading the model but it didn't work.

Then while trying the `glTF-Embedded` type of the model it imported correctly without any error.
After executing the `glTF-Embedded` type if I again try to execute the `glTF` type of the same model it imports smoothly without any error.

Tried this for the `Box` model too and the same thing happened.
@m-agour can you try out the `viz_gltf` on your side and share your experience?

Thanks!

File: fury/gltf.py [hunk 18]
```diff
@@ -0,0 +1,348 @@
...
+        for i, polydata in enumerate(self.polydatas):
+            actor = utils.get_actor_from_polydata(polydata)
+
+            if self.materials[i] is not None:
+                baseColorTexture = self.materials[i]['baseColorTexture']
+                actor.SetTexture(baseColorTexture)
+
+            self.actors.append(actor)
+
+        return self.actors
```
[the remark this answers is not in the document]
Comment (xtanion) [reply fury/gltf.py:69]: Polydata and materials are saved in the same order. In the `load_mesh` function if there's no material data for the primitive, It appends `None`.

So in this case (`[polydata_with_texture1, polydata_no_texture, polydata_with_texture2]`) the self.materials would be `[texture1, None, texture2]`.

**Comment by m-agour** [comment]
> Hello @xtanion , While I was trying to execute the `viz_gltf` file, I am facing some strange issues mentioned below.
> 
> * The fetcher is giving `RuntimeError: Event loop is closed` which will get solved when you merge the [Fixing API limits reached issue in gltf fetcher #616](https://github.com/fury-gl/fury/pull/616) as we discussed.
> * Also after successfully downloading the Model while importing, I am getting the below error
> 
> ```python
> Failed to read ! Error in opening file
> Traceback (most recent call last):
>   File "\docs\tutorials\01_introductory\viz_gltf.py", line 28, in <module>
>     gltf_obj = glTF(filename, apply_normals=True)
>   File "\fury\gltf.py", line 50, in __init__
>     self.get_nodes(0)
>   File "\fury\gltf.py", line 83, in get_nodes
>     self.transverse_node(node_id, self.init_transform)
>   File "\fury\gltf.py", line 129, in transverse_node
>     self.transverse_node(child_id, next_matrix)
>   File "\fury\gltf.py", line 121, in transverse_node
>     self.load_mesh(mesh_id, next_matrix)
>   File "\fury\gltf.py", line 148, in load_mesh
>     vertices = transform.apply_transfomation(vertices, transform_mat)
>   File "\fury\transform.py", line 335, in apply_transfomation
>     shape = vertices.shape
> AttributeError: 'NoneType' object has no attribute 'shape'
> ```
> 
> The vertices in the `apply_transformation` method is getting None. Also as you mentioned I tried redownloading the model but it didn't work.
> 
> Then while trying the `glTF-Embedded` type of the model it imported correctly without any error. After executing the `glTF-Embedded` type if I again try to execute the `glTF` type of the same model it imports smoothly without any error.
> 
> Tried this for the `Box` model too and the same thing happened. @m-agour can you try out the `viz_gltf` on your side and share your experience?
> 
> Thanks!

Yes @ganimtron-10, I'm getting the same error.

File: fury/gltf.py [hunk 19]
```diff
@@ -0,0 +1,348 @@
...
+        Path of the gltf file
+    apply_normals : bool, optional
+        If `True` applies normals to the mesh.
+    """
+
+    def __init__(self, filename, apply_normals=False):
+
+        self.gltf = GLTF2().load(filename)
+
+        gltf_path = filename.split('/')[-1:][0]
```
[the remark this answers is not in the document]
Comment (xtanion) [reply fury/gltf.py]: Hi @ganimtron-10, the issue you mentioned was caused by this line. I and @m-agour did some debugging and the using `os,path.basename` worked, as @skoudoro mentioned here.

Here we only need the base directory in `self.pwd` so I used `self.pwd = os.path.dirname(filename)` instead.

File: fury/gltf.py [hunk 20]
```diff
@@ -0,0 +1,348 @@
...
+            out_arr = np.frombuffer(buff_data, dtype=d_type,
+                                    count=byte_length, offset=byte_offset)
+
+            out_arr = out_arr.reshape(-1, byte_stride)
+            return out_arr
+
+        except IOError:
+            print('Failed to read ! Error in opening file')
+
+    def get_materials(self, mat_id):
```
[the remark this answers is not in the document]
Comment (xtanion) [reply fury/gltf.py:265]: Sorry for the doc, `get_materials` actually returns the dictionary of all textures. I modified the doc now.

**Comment by xtanion** [comment]
Hi @skoudoro, using `io.load_image` now, and added tests for new util functions, PTAL.
Thanks.

File: fury/gltf.py [hunk 21]
```diff
@@ -0,0 +1,341 @@
...
+                matnode = np.dot(matnode, scale)
+
+        next_matrix = np.dot(matrix, matnode)
+
+        if node.mesh is not None:
+            mesh_id = node.mesh
+            self.load_mesh(mesh_id, next_matrix)
+
+        if node.camera is not None:
+            camera_id = node.camera
```
Comment (skoudoro) [inline fury/gltf.py:124]: `camera_id` is not used.  is it on purpose or there is an error below?

**Review by skoudoro — COMMENTED** [review]
Hi @xtanion,

See the comment below. Thanks!

**Comment by skoudoro** [comment]
Also, When I run the tests locally, I got the following error:

```python
=============================================================================================== short test summary info ================================================================================================
FAILED fury/tests/test_gltf.py::test_load_gltf - TypeError: expected str, bytes or os.PathLike object, not NoneType
FAILED fury/tests/test_gltf.py::test_load_texture - TypeError: expected str, bytes or os.PathLike object, not NoneType
```

**Comment by xtanion** [comment]
Hi @skoudoro, are the tests failing because the mode is not explicitly mentioned in the `read_viz_gltf` function? Can you check it now?
Thanks

**Comment by xtanion** [comment]
Hi @skoudoro, I created [this](https://github.com/xtanion/fury/blob/gltf-mesh-camera/fury/gltf.py#L334) branch in which `load_camera` saves the camera as `vtkCamera` object. you can run the [viz_gltf.py](https://github.com/xtanion/fury/blob/gltf-mesh-camera/docs/tutorials/01_introductory/viz_gltf.py) model to see the diffrence. Please let me know If I should add that to this PR too.

**Comment by ganimtron-10** [comment]
Hello @xtanion ,
While importing various models I was facing the below issues

1. After importing the models they aren't clearly visible as you can see below (I don't know whether it is related to importing or light or something else)

    1. BoomBox:
BoomBox

    1. BarramundiFish:
BarramundiFish

1. Unicode❤♻Test:
It seems `pygltflib` is unable to decode the Unicode characters
```python
Traceback (most recent call last):
  File "d:\FuryWorkspace\fury\docs\tutorials\01_introductory\viz_gltf.py", line 28, in <module>
    gltf_obj = glTF(filename, apply_normals=True)
    self.gltf = GLTF2().load(filename)
  File "D:\FuryWorkspace\fury-env\lib\site-packages\pygltflib\__init__.py", line 1102, in load
    obj = cls.load_json(fname)
  File "D:\FuryWorkspace\fury-env\lib\site-packages\pygltflib\__init__.py", line 1050, in load_json
    obj = cls.gltf_from_json(f.read())
  File "C:\Users\ganim\AppData\Local\Programs\Python\Python39\lib\encodings\cp1252.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 285: character maps to <undefined>
```

1. Fox:
`glTF` type doesn't display any model
image
Do we support the `.glb` file??
Because when I try to import `Fox` in `glTF-Binary`  tmode it throws the below error
```python
Traceback (most recent call last):
  File "d:\FuryWorkspace\fury\docs\tutorials\01_introductory\viz_gltf.py", line 28, in <module>
    gltf_obj = glTF(filename, apply_normals=True)
  File "gltf.py", line 49, in __init__
    self.inspect_scene(0)
  File "gltf.py", line 82, in inspect_scene
    self.transverse_node(node_id, self.init_transform)
  File "gltf.py", line 120, in transverse_node
    self.load_mesh(mesh_id, next_matrix)
  File "gltf.py", line 146, in load_mesh
    vertices = self.get_acc_data(attributes.POSITION)
  File "gltf.py", line 209, in get_acc_data
    return self.get_buff_array(buff_id, d_type['dtype'], byte_length,
  File "gltf.py", line 247, in get_buff_array
    if uri.startswith('data:application/octet-stream;base64') or \
AttributeError: 'NoneType' object has no attribute 'startswith'
```
the `uri` is None. Yes, It is correct because the `.glb` file doesn't have any separate bin files for the buffer.

Also if we don't provide any mode to `read_viz_gltf`, the `glTF` mode isn't selected as the default.
This is because of [this line](https://github.com/fury-gl/fury/blob/master/fury/data/fetcher.py#L703) in `fetcher.py` where you are choosing the last type, so considering file order `glTF` will always be at the top (in windows as per I know) than the other  `glTF-*` type files.

I will continue to import another type of models too and let you know if I found something.

**Comment by skoudoro** [comment]
Thank you @xtanion for this work. There are minor issues to fix but let's do it on a new PR. 

Let's continue this integration! merging!

thanks a lot @ganimtron-10  and @m-agour for the great reviews.

**Pull request merged**